subsys/bluetooth/gatt_dm.c

```diff
@@ -61,69 +70,94 @@ struct bt_gatt_dm {
+	/* Round up len to 32 bits to make sure that return pointers are always
+	 * correctly aligned.
+	 */
+	len = (len + DATA_ALIGN - 1) & ~(DATA_ALIGN - 1);
 
-	if (!(dm->data_chunk.chunks[cur_chunk_id])) {
-		dm->data_chunk.chunks[cur_chunk_id] = k_malloc(CHUNK_SIZE);
-		dm->data_chunk.cur_len = 0;
+	__ASSERT_NO_MSG(len <= CHUNK_DATA_SIZE);
 
-		if (!(dm->data_chunk.chunks[cur_chunk_id])) {
-			return NULL;
-		}
-	}
-
-	size_t misalign = dm->data_chunk.cur_len % UUID_ALIGN;
-	if (misalign) {
-		dm->data_chunk.cur_len += UUID_ALIGN - misalign;
-	}
+	if (!dm->data_chunk.cur
+			|| dm->data_chunk.cur_len + len > CHUNK_DATA_SIZE) {
 
-	if (dm->data_chunk.cur_len + len > CHUNK_SIZE) {
-		cur_chunk_id++;
-		dm->data_chunk.cur_id = cur_chunk_id;
-		if ((dm->data_chunk.chunks[cur_chunk_id]) ||
-		    (cur_chunk_id >= ARRAY_SIZE(dm->data_chunk.chunks))) {
-			LOG_ERR("Not enough data chunks");
+		new_item = k_malloc(sizeof(struct data_chunk_item));
+		if (!new_item) {
 			return NULL;
 		}
 
-		return user_data_store(dm, user_data, len);
+		if (dm->data_chunk.cur) {
+			dm->data_chunk.cur->next = new_item;
+		} else {
+			dm->data_chunk.first = new_item;
+		}
+		dm->data_chunk.cur = new_item;
+		dm->data_chunk.cur->next = NULL;
+		dm->data_chunk.cur_len = 0;
 	}
 
-	user_data_loc =
-		&dm->data_chunk.chunks[cur_chunk_id][dm->data_chunk.cur_len];
-	memcpy(user_data_loc, user_data, len);
+	user_data_loc = &dm->data_chunk.cur->data[dm->data_chunk.cur_len];
 	dm->data_chunk.cur_len += len;
 
 	return user_data_loc;
 }
 
 static void svc_attr_memory_release(struct bt_gatt_dm *dm)
 {
+	struct data_chunk_item *cur_chunk;
+
 	LOG_DBG("Attr memory release");
 	/* Clear attributes */
-	memset(dm->attrs, 0, sizeof(dm->attrs));
 	dm->cur_attr_id = 0;
 	/* Release dynamic memory data chunks */
-	for (size_t i = 0; i <= dm->data_chunk.cur_id; i++) {
-		k_free(dm->data_chunk.chunks[i]);
-		dm->data_chunk.chunks[i] = NULL;
+	while (dm->data_chunk.first) {
+		cur_chunk = dm->data_chunk.first;
+		dm->data_chunk.first = cur_chunk->next;
+		k_free(cur_chunk);
 	}
-	dm->data_chunk.cur_id = 0;
+	dm->data_chunk.cur = NULL;
 	dm->data_chunk.cur_len = 0;
 }
 
-static struct bt_gatt_attr *attr_store(struct bt_gatt_dm *dm,
-				       const struct bt_gatt_attr *attr)
+/* Returns size of UUID structure with padding for memory alignment */
+static size_t get_uuid_size(const struct bt_uuid *uuid)
+{
+	switch (uuid->type) {
+	case BT_UUID_TYPE_16:
+		return sizeof(struct bt_uuid_16);
+	case BT_UUID_TYPE_32:
+		return sizeof(struct bt_uuid_32);
+	case BT_UUID_TYPE_128:
+		return sizeof(struct bt_uuid_128);
+	default:
+		LOG_ERR("Unsupported UUID type.");
+		return 0;
+	}
+}
+
+/** @brief Stores attribute in bt_gatt_dm instance.
+ *
+ * This function stores attr at dm->attrs array. Its UUID is stored in
+ * dm->data_chunk. Additional buffer can be allocated before UUID. It is used
+ * to store bt_gatt_service_val or bt_gatt_chrc associated with this attr.
```

Comment:
It took me a while to understand why we need to allocate the space (uuid size + additional_len) in one go.

This function stores attr at dm->attrs array. Its UUID is stored in dm->data_chunk.
[new/changed description starts here]
The DM attribute descriptor does not contain a pointer to the context data. This data could be either bt_gatt_service_val or bt_gatt_chrc. It is assumed that attribute context data (if any) is always placed before its UUID data. For this purpose, an additional buffer is allocated and used.

Reply:
Done